content/en/blog/2022-03-11-flagger-is-adding-gateway-api/index.md

```diff
@@ -0,0 +1,213 @@
+
+![Gateway API Model](api-model.png)
+
+Gateway API exposes a more general API than Ingress for proxying and you
+can use it for more protocols than just HTTP (although most
+implementations support just HTTP for now). It models more
+infrastructure components to provide better deployment and management
+options. There are three core components to the Gateway API:
+
+- `GatewayClass`: This lets us define which controller implementation we
+  want to use.
+- `Gateway`: A `Gateway` resource is attached to a `GatewayClass` and
+  has a 1:1 relationship with the actual load balancing infra. It lets
+  us define a set of listeners, through which we can specify which
+  Route resources to evaluate for routing, amongst other things.
+- `HTTPRoute`: This is a `Route` resource that is specific for HTTP
+  requests. It defines routing rules such as filters, path and
+  header matches, etc. and which services should the request be
+  forwarded to.
+
+## How does this work in Flagger?
+
+Flagger makes use of the fact that `HTTPRoute` allows users to define a
+weight related to each routing rule. These weights are used to determine
+which service should receive a request. For example, if we want to send
+10% of our traffic to another service, we can define a `HTTPRoute` like:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: foo-route
+spec:
+  parentRefs:
+  - name: example-gateway
+  hostnames:
+  - "foo.example.com"
+  rules:
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /login
+    backendRefs:
+    - name: foo-svc
+      port: 8080
+      weight: 90
+    - name: foo-svc-canary
+      port: 8080
+      weight: 10
+```
+
+This sends 10% of all requests coming to foo.example.com/login to the
+new service and the other 90% requests go to the stable service. You can
+read more about traffic splitting in Gateway API
+[here](https://gateway-api.sigs.k8s.io/v1alpha2/guides/traffic-splitting/).
+
+Flagger creates a `HTTPRoute` with the appropriate header matches, path
+matches, etc and attaches the primary and canary service to the
+`HTTPRoute`. During the canary analysis, the weights related to both the
+services are adjusted accordingly.
+
+![Flagger using Gateway API](featured-flagger-gatewayapi-canary.png)
+
+If you want to get started right away, have a look at [our
+tutorial](https://docs.flagger.app/tutorials/gatewayapi-progressive-delivery),
+which shows you how to use Gateway API and Flagger to automate canary
+deployments. It won't take long to follow, but will convey how powerful
+this integration is.
+
+## Flagger works with all implementations
+
+With added support for Gateway API, Flagger now works with all
+[implementations](https://gateway-api.sigs.k8s.io/implementations/),
+which means that as of today these are natively supported: Apache
+APISIX, Contour, Emissary-Ingress, Gloo Edge 2.0, Google Kubernetes
+Engine, HAProxy Ingress, HashiCorp Consul, Istio, Kong and Traefik.
+
```

Comment:
The Flagger team has successfully tested Contour and Istio using the v1beta2 Gateway API.  Starting with Flagger v1.19, the Gateway API is part of our end-to-end test suite using the Contour implementation.